src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts

```diff
@@ -360,16 +419,26 @@ export class ObjectExplorerService implements IObjectExplorerService {
 	public async createNewSession(providerId: string, connection: ConnectionProfile): Promise<azdata.ObjectExplorerSessionResponse> {
 		const provider = this._providers[providerId];
 		if (provider) {
+			// set the connection to wait for session
+			this._connectionsWaitingForSession.set(connection.id, true);
 			const result = await provider.createNewSession(connection.toConnectionInfo());
+			// some providers return a malformed create sessions responses which don't have a session id. We should throw an error in this case
+			if (!result?.sessionId) {
+				this.logService.error(`The session ID returned by provider "${providerId}" for connection "${connection.title}" is invalid.`);
+				throw new Error(nls.localize('objectExplorerSessionIdMissing', 'The session ID returned by provider "{0}" for connection "{1}" is invalid.', providerId, connection.title));
+			}
 			if (this._sessions[result.sessionId]) {
 				this.logService.trace(`Overwriting session ${result.sessionId}`);
 			}
 			this._sessions[result.sessionId] = {
 				connection: connection,
 				nodes: {}
 			};
+			// once the session is created, set the connection to not wait for session
+			this._connectionsWaitingForSession.set(connection.id, false);
```

Comment:
If you store a deferred promise here instead of just a boolean then you can use that to complete the promise above in place of the polling timeout (or use an event)